akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -171,35 +383,46 @@ private[akka] class Shard(
     settings: ClusterShardingSettings,
     extractEntityId: ShardRegion.ExtractEntityId,
     @unused extractShardId: ShardRegion.ExtractShardId,
-    handOffStopMessage: Any)
+    handOffStopMessage: Any,
+    rememberEntitiesProvider: Option[RememberEntitiesProvider])
     extends Actor
     with ActorLogging
+    with Stash
     with Timers {
 
   import Shard._
   import ShardCoordinator.Internal.HandOff
   import ShardCoordinator.Internal.ShardStopped
   import ShardRegion.EntityId
-  import ShardRegion.Msg
   import ShardRegion.Passivate
   import ShardRegion.ShardInitialized
   import ShardRegion.handOffStopperProps
-  import settings.tuningParameters._
 
   import akka.cluster.sharding.ShardCoordinator.Internal.CoordinatorMessage
-  import akka.cluster.sharding.ShardRegion.ShardRegionCommand
 
-  var state = State.Empty
-  var idByRef = Map.empty[ActorRef, EntityId]
-  var refById = Map.empty[EntityId, ActorRef]
-  var lastMessageTimestamp = Map.empty[EntityId, Long]
-  var passivating = Set.empty[ActorRef]
-  val messageBuffers = new MessageBufferMap[EntityId]
+  final val verboseDebug = context.system.settings.config.getBoolean("akka.cluster.sharding.verbose-debug-logging")
+
+  private val rememberEntitiesStore: Option[ActorRef] =
+    rememberEntitiesProvider.map { provider =>
+      val store = context.actorOf(provider.shardStoreProps(shardId).withDeploy(Deploy.local), "RememberEntitiesStore")
+      context.watchWith(store, RememberEntityStoreCrashed(store))
+      store
+    }
+
+  private val flightRecorder = ShardingFlightRecorder(context.system)
+
+  private val entities = new Entities(log, settings.rememberEntities, verboseDebug)
+
+  private var lastMessageTimestamp = Map.empty[EntityId, Long]
```

Comment:
should this also move into `Entities`?

Reply:
Not sure, it's not strictly related to the aggregate Entities state, if we could move it into Active that could make sense, since it is for a specific entity, but then we'd need to scan all entities to find which ones to passivate, so maybe leave it here.

Reply:
ok